crates/c-api/include/wasmtime/component.h

```diff
@@ -0,0 +1,635 @@
+ * \param results where to write the results of this function
+ * \param nresults how many results must be produced
+ *
+ * Callbacks are guaranteed to get called with the right types of arguments, but
+ * they must produce the correct number and types of results. Failure to do so
+ * will cause traps to get raised on the wasm side.
+ *
+ * This callback can optionally return a #wasm_trap_t indicating that a trap
+ * should be raised in WebAssembly. It's expected that in this case the caller
+ * relinquishes ownership of the trap and it is passed back to the engine.
+ */
+typedef wasm_trap_t *(*wasmtime_component_func_callback_t)(
+    void *env, wasmtime_context_t *context,
+    const wasmtime_component_val_t *args, size_t nargs,
+    wasmtime_component_val_t *results, size_t nresults);
+
+/**
+ * \brief Defines a host function in a linker
+ *
+ * Must be done before calling #wasmtime_component_linker_build
+ *
+ * Does not take ownership of the #wasmtime_component_type_t arguments.
+ *
+ * \param linker the #wasmtime_component_linker_t in which the function should
+ *        be defined
+ * \param path the dot-separated path of the package where the function is
+ *        defined
+ * \param path_len the byte length of `path`
+ * \param name the name of the function
+ * \param name_len the byte length of `name`
+ * \param params_types_buf a pointer to an array of #wasmtime_component_type_t
+ *        describing the function's parameters
+ * \param params_types_len the length of `params_types_buf`
+ * \param outputs_types_buf a pointer to an array of #wasmtime_component_type_t
+ *        describing the function's outputs
+ * \param outputs_types_len the length of `outputs_types_buf`
+ * \param cb a pointer to the actual functions, a
+ *        #wasmtime_component_func_callback_t
+ * \param data the host-provided data to provide as the first argument to the
+ *        callback
+ * \param finalizer an optional finalizer for the `data` argument.
+ *
+ * \return wasmtime_error_t* on success `NULL` is returned, otherwise an error
+ *         is returned which describes why the definition failed.
+ */
+wasmtime_error_t *wasmtime_component_linker_define_func(
+    wasmtime_component_linker_t *linker, const char *path, size_t path_len,
+    const char *name, size_t name_len,
+    wasmtime_component_type_t *params_types_buf, size_t params_types_len,
+    wasmtime_component_type_t *outputs_types_buf, size_t outputs_types_len,
+    wasmtime_component_func_callback_t cb, void *data,
+    void (*finalizer)(void *));
+
+/**
+ * \brief Builds the linker, providing the host functions defined by calls to
+ * #wasmtime_component_linker_define_func
+ *
+ * \param linker the #wasmtime_component_linker_t to build
+ *
+ * \return wasmtime_error_t* On success `NULL` is returned, otherwise an error
+ *         is returned which describes why the build failed.
+ */
+wasmtime_error_t *
+wasmtime_component_linker_build(wasmtime_component_linker_t *linker);
```

Comment:
This was the way I tried to "hide" the difference between `Linker` and `LinkerInstance`, the latter didn't seem useful to expose to the C api. But in order to add host functions to the linker, you need the appropriate `LinkerInstance`, and that can only be created once (through `.root()` of `.instance(name)`). An alternative could be to keep the `LinkerInstances` around in a map, or to provide a lookup mechanism to `Linker`.

Reply:
FWIW I think it's fine to change the Rust APIs to better accomodate the C APIs if necessary, but otherwise I'd prefer to keep the two relatively in-sync to make them more managable to maintain, so having a `LinkerInstance` equivalent in the C API I think would be reasonable (or changing the Rust API to not be based on `LinkerInstance`)